pkg/mcp/client.go

```diff
@@ -0,0 +1,365 @@
+
+package mcp
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	mcpclient "github.com/mark3labs/mcp-go/client"
+	mcp "github.com/mark3labs/mcp-go/mcp"
+	"k8s.io/klog/v2"
+)
+
+// Client represents an MCP client that can connect to MCP servers
+type Client struct {
+	// Name is a friendly name for this MCP server connection
+	Name string
+	// Command is the command to execute for stdio-based MCP servers
+	Command string
+	// Args are the arguments to pass to the command
+	Args []string
+	// Env are the environment variables to set for the command
+	Env []string
+	// client is the underlying MCP client
+	client *mcpclient.Client
+	// Note: cmd field removed since NewStdioMCPClient handles the server process automatically
+}
+
+// NewClient creates a new MCP client with the given configuration
+func NewClient(name, command string, args []string, env map[string]string) *Client {
+	// Convert env map to slice of KEY=value strings
+	var envSlice []string
+	for k, v := range env {
+		envSlice = append(envSlice, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	return &Client{
+		Name:    name,
+		Command: command,
+		Args:    args,
+		Env:     envSlice,
+	}
+}
+
+// Connect establishes a connection to the MCP server
+// It creates an MCP client that will start the server process automatically
+func (c *Client) Connect(ctx context.Context) error {
+	klog.V(2).InfoS("Connecting to MCP server", "name", c.Name, "command", c.Command, "args", c.Args)
+	if c.client != nil {
+		return nil // Already connected
+	}
+
+	// Step 1: Prepare environment and command
+	expandedCmd, finalEnv, err := c.prepareEnvironment()
+	if err != nil {
+		return fmt.Errorf("preparing environment: %w", err)
+	}
+
+	// Step 2: Create the MCP client
+	if err := c.createMCPClient(expandedCmd, finalEnv); err != nil {
+		return fmt.Errorf("creating MCP client: %w", err)
+	}
+
+	// Step 3: Initialize the connection
+	if err := c.initializeConnection(ctx); err != nil {
+		c.cleanup()
+		return fmt.Errorf("initializing connection: %w", err)
+	}
+
+	// Step 4: Verify the connection
+	if err := c.verifyConnection(ctx); err != nil {
+		c.cleanup()
+		return fmt.Errorf("verifying connection: %w", err)
+	}
+
+	klog.V(2).Info("Successfully connected to MCP server", "name", c.Name)
+	return nil
+}
+
+// prepareEnvironment expands the command path and merges environment variables
+func (c *Client) prepareEnvironment() (string, []string, error) {
+	// Expand the command path to handle ~ and environment variables
+	expandedCmd, err := expandPath(c.Command)
+	if err != nil {
+		return "", nil, fmt.Errorf("expanding command path: %w", err)
+	}
+
+	// Build proper environment slice by merging process environment with custom env
+	finalEnv := mergeEnvironmentVariables(os.Environ(), c.Env)
+
+	return expandedCmd, finalEnv, nil
+}
+
+// createMCPClient creates the underlying MCP client
+func (c *Client) createMCPClient(command string, env []string) error {
+	client, err := mcpclient.NewStdioMCPClient(command, env, c.Args...)
```

Comment:
Aah...doesn't look like we are supporting remote mcp client... right ?

Reply:
You’re right. I overlooked this. I’ll add a `TODO` for adding SSE support.

Reply:
You’re right. I overlooked this. I’ll add a `TODO` for adding SSE support.

Reply:
yes, stdio only for now. Please see comment above.